runtime/src/accounts.rs

```diff
@@ -136,13 +144,26 @@ impl Accounts {
             // If a fee can pay for execution then the program will be scheduled
             let mut called_accounts: Vec<Account> = vec![];
             let mut credits: InstructionCredits = vec![];
-            for key in &message.account_keys {
+            for (i, key) in message.account_keys.iter().enumerate() {
                 if !message.program_ids().contains(&key) {
-                    called_accounts.push(
-                        AccountsDB::load(storage, ancestors, accounts_index, key)
-                            .map(|(account, _)| account)
-                            .unwrap_or_default(),
-                    );
+                    let account = AccountsDB::load(storage, ancestors, accounts_index, key)
+                        .map(|(account, _)| account)
+                        .unwrap_or_default();
+                    if !message.is_debitable(i) {
+                        let position = message
+                            .get_account_keys_by_lock_type()
+                            .1
+                            .iter()
+                            .position(|k| k == &key)
+                            .unwrap();
+                        if credit_only_locks[position].load(Ordering::Relaxed) == 0
+                            && account.lamports > 0
+                        {
+                            credit_only_locks[position].store(account.lamports, Ordering::Relaxed);
+                        }
+                    }
+
+                    called_accounts.push(account);
```

Comment:
Hmmm, so even if the account is a credit-only account, we push the stored account from AccountsDb instead of the cached value in the `credit_only_locks`?

Reply:
Yes. This is the part that actually protects against the issue you describe in your comment below, @carllin . Until credits are actually stored to the credit-only account (at the end of the slot, or when only one refcount is held, meaning there are no other read attempts occuring), all account loads read the original pre-credit account. Then, all credits are stored to the account in one action, so subsequent reads will all see the same balance again.
Does that make sense?

Reply:
Hmmm the problem is entry ordering. Separate Entries should be able to run in parallel with each other and lead to deterministic results. However, it seems possible in the current design to create two separate entries (imagine in the example below that thread 1 creates an entry and thread 2 creates an entry), where depending on order of execution of those two entries, leads to different results

Reply:
The only place I can see that being possible is if the read (thread 1) begins after this line executes on thread 2 (https://github.com/solana-labs/solana/pull/4691/files#diff-0104dfa96426a3a49428815872a0356eR509), but before the subsequent `accounts_db.store` completes. Do you concur?
That does argue for some sort of lock on the account during the moments the actual store is occurring, or (the perhaps simpler solution) to only store credit-only accounts at the end of a slot, when we call freeze()